src/framework/ui/internal/uiengine.cpp

```diff
@@ -107,17 +107,27 @@ void UiEngine::setup(QQmlEngine* engine)
 
     m_engine = engine;
     m_theme->init();
-    m_engine->rootContext()->setContextProperty("ui", this);
+    QJSValue ui = m_engine->newQObject(this);
+    QQmlEngine::setObjectOwnership(this, QQmlEngine::CppOwnership);
+    m_engine->globalObject().setProperty("ui", ui);
+
     m_engine->rootContext()->setContextProperty("api", m_api);
 
     QJSValue translator = m_engine->newQObject(m_translation);
     QJSValue translateFn = translator.property("translate");
-    m_engine->globalObject().setProperty("qsTrc", translateFn);
+    m_engine->setObjectOwnership(m_translation, QQmlEngine::CppOwnership);
+    m_translationFunction = QVariant::fromValue(translateFn);
+
+    QJSValue qsTrc = m_engine->evaluate(
+        "(function(context, text, disambiguation = \"\", n = -1) {"
+        "    return ui.trc(context, text, disambiguation, n);"
+        "})");
```

Comment:
I also thought about something similar, but I didn't know that we can call properties with arguments, it's cool!